README.md
# ECE 198 Project - Hot Drink Temperature Tracker

This project is for the ECE 198 Project Studio course in 1A Computer Engineering at the University of Waterloo, created by Wilson Cheng and William Zhang.

Its purpose is to track the temprature of a hot mug, and ring a buzzer when it has reached a preset desired temperature. 

All this was done by programming an STM32F401RE with a breadboard, resistors, wires, and the peripherals mentioned below. 

## Getting started

To utilize this, follow the documents in the "Design Files" directory to assemble the circuits and systems.

Then, to flash the code, plug the STM32 into an electronic source, clone this repository, open an STM CubeIDE, and press the build hammer and the run button.

The display should prompt you to press the button, after which the thermometer will take an initial measurement. Then, you should be prompted to turn the potentiometer knob to set a temperature. Push the button again to set your desired temperature, and wait for the buzz. You will see regular temperature updates. 

## How does this work?

The project will involve the control of 5 peripherals: an LCD display, an infrared thermometer, a push button, a buzzer, and a potentiometer.

This is accomplished with GPIO input/output control, ADC input, and the I2C communication protocol. 

### How is each peripheral controlled?

1) The display: we used a library whose code is in the files called `liquidcrystal_i2c`. This allowed for easy calling of functions, after setting the I2C pins, checking the `hi2c1`, and generating code. 

2) The thermometer: another library was used, whose code is in [this](https://github.com/dinamitemic/mlx90614) repo. However, certain functions were uncallable with our STM32, and external variables had to be modified to `hi2c1` rather than the original `hi2c3`. Finally, while the thermometer can detect ambient temperature if used with library functions on Arduino, there were no ambient temperature functions here. Only object temperature functions, which had to be modified, and in fact, extracted to `main.c` to make temperature detection work.

3) The potentiometer: here basic ADC was used. A potentiometer works by splitting the input voltage into two. The potential difference between the outer two pins remains the same, however, turning the knob will adjust the resistance and the output voltage in the inner pin (connected to a pin set to `ADC`). The potentiometer used in this project was linear, so we used a linear interpolation equation allowing adjustment on integer values from 20 degrees (standard room temperature) and the initial measurement, with single degree increments. See main.c comments.

4) The buzzer: this was a very simple peripheral, with no resistors needed, only a GPIO output, and a ground pin. When it was time to buzz, the GPIO output was set, and kept on for half a second with `HAL_Delay(500)` until the pin was reset. 

5) The push button: this peripheral was slightly complex, as the pins had to be bent straight to stick all the way in the breadboard for the button to work! The way this was set up was through a 5V to Ground connection with a 10k Ohm pull-up resistor in between, and an input pin branched out to the STM32 microcontroller. If the input pin detected a 0 state, which indicates that the button is pressed down, actions were taken accordingly, depending on the stage of the procedure.

### How is this all put together?

We use the power of enumerations! In the C programming language, enums are user-defined data types with user-defined readable names for the set of possible constants. 

With firmware and embedded systems programming, after you flash the code to the microcontroller, the microcontroller will continuously run the `while(1) {}` loop. For the special case of the STM32, when you open up the CubeIDE, you choose a board type, for which code is immediately set up by the IDE for that board and more code can be generated. 

Then, there are "user code" comments, between which one should write code to preserve it through code regenerations. To set pins to a certain state and configure them, one simply opens the `.ioc` file eponymous with the user-defined project name, sets configurations, saves the `.ioc` file, and the CubeIDE will generate code for you to configure everything.

**Therefore**, this project will have to run continuous, possibly infinite, rounds of the "measure-set temperature-check temperature-buzz" steps. In each step, the same while loop must be continuously traversed to monitor user input. So with different sets of sets of steps in the same loop, we simply associate enum constants such as `reset`, `set-temp`, etc with user-defined enums constants and certain commands via control flow. Side joke: ChatGPT was not used for that.

## Further Notes

An extra feature that was added to stop infinite runs is to check if the measured temperature is lower than the desired temperature. This occurs outside of existing conditions when both the desired temperature and cup temperature are much higher than room temperature, and the cup is removed before the buzzer sounds. 

The design files include the design document, the circuit diagrams, the 3D model of the stand. Note that the changelog markdown stores all version changes. 

To view a demonstration of a successful implementation, see the folder called 'Demo Videos'. Apologies for the poor filming quality and display resolution.

CHANGELOG.md
# CHANGELOG

Documentation of changes to the project. 

## [v.0.3] - 2023-11-19

## Added:
- new redundancy feature to check if measured temperature is below desired temperature
- merged to main as last revision in `a234316e`

## [v.0.2] - 2023-11-19

### Added:
- A 1 second delay after final temperature measurement so user can verify that it is close to desired temperature.
- merged to main in `b8ecdb88`

### Changed:
- Changed from 30 second delay to 10 second delay

## [v.0.1] - 2023-11-16

### Added:
- *Mechanical*: new stand configuration, from the file added in commit `07b0906e`. It has a cardboard guard rail, and a box to lower the thermometer.

### Changed:
- fixed potentiometer linear interpolation formula to T = ((ADC_Value)*(init_temp-20)/4095)+20, where 
    - T is temperature mapped by potentiometer
    - init_temp is initial temperature measurement
    - ADC_Value is raw potentiometer value received by ADC 
- merged to main in `673a10b9`

## [v.0.0] - 2023-11-16

### Added:

- Made procedure work according to original design
- Configured loops and delays
- In commit `f562d5b0`
- *Schematic*: two significant feature changes in commit `514a44f2`. This includes:
    - 10k Ohm Resistors added to SDA/SCL connection with 5V
    - Voltage and GPIO pin changes on button and potentiometer
- *Mechanical*: stand configuration with rail according to original design, in commit `cbe37fea`

Core/Src/mlx90614.c
/*
 * mlx90614.c
 *
 *  Created on: Nov 11, 2023
 *      Author: wzhan
 */

#include <assert.h>
#include <stdio.h>
#include <stdint.h>
#include <string.h>
#include "stm32f4xx_hal.h"
#include "mlx90614.h"

extern I2C_HandleTypeDef hi2c1;
extern UART_HandleTypeDef huart2;
uint8_t temp_buff[128] = {};

static const uint8_t crc_table[] = {
    0x00, 0x07, 0x0e, 0x09, 0x1c, 0x1b, 0x12, 0x15, 0x38, 0x3f, 0x36, 0x31,
    0x24, 0x23, 0x2a, 0x2d, 0x70, 0x77, 0x7e, 0x79, 0x6c, 0x6b, 0x62, 0x65,
    0x48, 0x4f, 0x46, 0x41, 0x54, 0x53, 0x5a, 0x5d, 0xe0, 0xe7, 0xee, 0xe9,
    0xfc, 0xfb, 0xf2, 0xf5, 0xd8, 0xdf, 0xd6, 0xd1, 0xc4, 0xc3, 0xca, 0xcd,
    0x90, 0x97, 0x9e, 0x99, 0x8c, 0x8b, 0x82, 0x85, 0xa8, 0xaf, 0xa6, 0xa1,
    0xb4, 0xb3, 0xba, 0xbd, 0xc7, 0xc0, 0xc9, 0xce, 0xdb, 0xdc, 0xd5, 0xd2,
    0xff, 0xf8, 0xf1, 0xf6, 0xe3, 0xe4, 0xed, 0xea, 0xb7, 0xb0, 0xb9, 0xbe,
    0xab, 0xac, 0xa5, 0xa2, 0x8f, 0x88, 0x81, 0x86, 0x93, 0x94, 0x9d, 0x9a,
    0x27, 0x20, 0x29, 0x2e, 0x3b, 0x3c, 0x35, 0x32, 0x1f, 0x18, 0x11, 0x16,
    0x03, 0x04, 0x0d, 0x0a, 0x57, 0x50, 0x59, 0x5e, 0x4b, 0x4c, 0x45, 0x42,
    0x6f, 0x68, 0x61, 0x66, 0x73, 0x74, 0x7d, 0x7a, 0x89, 0x8e, 0x87, 0x80,
    0x95, 0x92, 0x9b, 0x9c, 0xb1, 0xb6, 0xbf, 0xb8, 0xad, 0xaa, 0xa3, 0xa4,
    0xf9, 0xfe, 0xf7, 0xf0, 0xe5, 0xe2, 0xeb, 0xec, 0xc1, 0xc6, 0xcf, 0xc8,
    0xdd, 0xda, 0xd3, 0xd4, 0x69, 0x6e, 0x67, 0x60, 0x75, 0x72, 0x7b, 0x7c,
    0x51, 0x56, 0x5f, 0x58, 0x4d, 0x4a, 0x43, 0x44, 0x19, 0x1e, 0x17, 0x10,
    0x05, 0x02, 0x0b, 0x0c, 0x21, 0x26, 0x2f, 0x28, 0x3d, 0x3a, 0x33, 0x34,
    0x4e, 0x49, 0x40, 0x47, 0x52, 0x55, 0x5c, 0x5b, 0x76, 0x71, 0x78, 0x7f,
    0x6a, 0x6d, 0x64, 0x63, 0x3e, 0x39, 0x30, 0x37, 0x22, 0x25, 0x2c, 0x2b,
    0x06, 0x01, 0x08, 0x0f, 0x1a, 0x1d, 0x14, 0x13, 0xae, 0xa9, 0xa0, 0xa7,
    0xb2, 0xb5, 0xbc, 0xbb, 0x96, 0x91, 0x98, 0x9f, 0x8a, 0x8d, 0x84, 0x83,
    0xde, 0xd9, 0xd0, 0xd7, 0xc2, 0xc5, 0xcc, 0xcb, 0xe6, 0xe1, 0xe8, 0xef,
    0xfa, 0xfd, 0xf4, 0xf3
};

uint8_t CRC8_Calc (uint8_t *p, uint8_t len) {
        uint16_t i;
        uint16_t crc = 0x0;

        while (len--) {
                i = (crc ^ *p++) & 0xFF;
                crc = (crc_table[i] ^ (crc << 8)) & 0xFF;
        }

        return crc & 0xFF;
}

void MLX90614_WriteReg(uint8_t devAddr, uint8_t regAddr, uint16_t data) {

	uint8_t i2cdata[4], temp[4];

	temp[0] = (devAddr << 1);
	temp[1] = regAddr;

	temp[2] = 0x00;
	temp[3] = 0x00;

	// For a write word command, in the crc8 calculus, you have to include [SA_W, Command, LSB, MSB]

	i2cdata[0] = temp[1]; //EEPROM-address
	i2cdata[1] = temp[2]; //Delete-Byte, low
	i2cdata[2] = temp[3]; //Delete-Byte, high
	i2cdata[3] = CRC8_Calc(temp, 4); //CRC8-checksum calculation: http://www.sunshine2k.de/coding/javascript/crc/crc_js.html

	HAL_I2C_Master_Transmit(&hi2c1, (devAddr << 1), i2cdata, 4, 0xFFFF);
	HAL_Delay(10);

	MLX90614_SendDebugMsg(MLX90614_DBG_MSG_W, devAddr, i2cdata[0], (i2cdata[1] <<8 | i2cdata[2]), i2cdata[3], 0x00);

	temp[2] = data & 0xFF; //Getting LSB first
	temp[3] = data >> 8;   //Getting MSB after

	i2cdata[0] = temp[1]; //EEPROM-address
	i2cdata[1] = temp[2]; //Delete-Byte, low
	i2cdata[2] = temp[3]; //Delete-Byte, high
	i2cdata[3] = CRC8_Calc(temp, 4); //CRC8-checksum calculation: http://www.sunshine2k.de/coding/javascript/crc/crc_js.html

	HAL_I2C_Master_Transmit(&hi2c1, (devAddr << 1), i2cdata, 4, 0xFFFF);
	HAL_Delay(10);
	MLX90614_SendDebugMsg(MLX90614_DBG_MSG_W, devAddr, i2cdata[0], data, i2cdata[3], 0x00);
}
int MLX90614_ReadReg(uint8_t devAddr, uint8_t regAddr, uint8_t dbg_lvl) {
	char i2cdata[2];

	HAL_I2C_Mem_Read(&hi2c1, 0xB4, 0x07, 1, (uint8_t*)i2cdata, 2, 100);

	// For a read word command, in the crc8 calculus, you have to include [SA_W, Command, SA_R, LSB, MSB]
	/*crc_buff[0] = (devAddr<<1);
	crc_buff[1] = regAddr;
	crc_buff[2] = (devAddr<<1) + 1;
	crc_buff[3] = in_buff[0];
	crc_buff[4] = in_buff[1];
	crc = CRC8_Calc(crc_buff, 5);*/

	int raw_temp = ((i2cdata[1] << 8) | (i2cdata[0]));

	//TODO: implement CRC8 check on data received
	/*if (crc != in_buff[2]) {
		data = 0x0000;
	}
	if(dbg_lvl == MLX90614_DBG_ON)	MLX90614_SendDebugMsg(MLX90614_DBG_MSG_R, devAddr, regAddr, data, in_buff[2], crc);
	 */
	return raw_temp;
}
float MLX90614_ReadTemp(uint8_t devAddr, uint8_t regAddr) { // dev addr must be sensor addr
	float temp;
	int raw_temp;

	raw_temp = MLX90614_ReadReg(devAddr, regAddr, MLX90614_DBG_OFF);
	temp = raw_temp*0.02 - 273.15;
	HAL_Delay(10);
	return temp;
}
void MLX90614_ScanDevices (void) {
	HAL_StatusTypeDef result;
	for (int i = 0; i<128; i++)
		  {
			  result = HAL_I2C_IsDeviceReady(&hi2c1, (uint16_t) (i<<1), 2, 2); // NOTE: actually checks the address shifted left
			  if (result != HAL_OK)
			  {
				  sprintf((char*)temp_buff, ".");
				  HAL_UART_Transmit(&huart2, temp_buff, strlen((const char *)temp_buff), HAL_MAX_DELAY);
				  HAL_Delay(500);
			  }
			  if (result == HAL_OK)
			  {
				  sprintf((char*)temp_buff, "0x%X", i);
				  HAL_UART_Transmit(&huart2, temp_buff, strlen((const char *)temp_buff), HAL_MAX_DELAY);
				  HAL_Delay(500);
			  }
		  }
}
void MLX90614_SendDebugMsg(uint8_t op_type, uint8_t devAddr, uint8_t regAddr, uint16_t data, uint8_t crc_in, uint8_t crc_calc) {
	if(op_type == MLX90614_DBG_MSG_W) {
		snprintf((char*)temp_buff, sizeof(temp_buff), "W Dev: 0x%02X, Reg: 0x%02X, Data: 0x%04X, CRC8_calc:0x%02X\r\n", devAddr, regAddr, data, crc_calc);
		HAL_UART_Transmit(&huart2, temp_buff, strlen((const char *)temp_buff), HAL_MAX_DELAY);
		HAL_Delay(500);
	}
	else if (op_type == MLX90614_DBG_MSG_R) {
		snprintf((char*)temp_buff, sizeof(temp_buff), "R Dev: 0x%02X, Reg: 0x%02X, Data: 0x%04X, CRC8_in:0x%02X, CRC8_calc:0x%02X\r\n", devAddr, regAddr, data, crc_in, crc_calc);
		HAL_UART_Transmit(&huart2, temp_buff, strlen((const char *)temp_buff), HAL_MAX_DELAY);
		HAL_Delay(500);
	}

}


Core/Src/main.c
/* USER CODE BEGIN Header */
/**
  ******************************************************************************
  * @file           : main.c
  * @brief          : Main program body
  ******************************************************************************
  * @attention
  *
  * Copyright (c) 2023 STMicroelectronics.
  * All rights reserved.
  *
  * This software is licensed under terms that can be found in the LICENSE file
  * in the root directory of this software component.
  * If no LICENSE file comes with this software, it is provided AS-IS.
  *
  ******************************************************************************
  */
/* USER CODE END Header */
/* Includes ------------------------------------------------------------------*/
#include "main.h"

/* Private includes ----------------------------------------------------------*/
/* USER CODE BEGIN Includes */
#include <string.h>
#include <stdbool.h>
#include <stdio.h>
#include <float.h>
#include <stdio.h>
#include <stdlib.h>
/* USER CODE END Includes */

/* Private typedef -----------------------------------------------------------*/
/* USER CODE BEGIN PTD */

/* USER CODE END PTD */

/* Private define ------------------------------------------------------------*/
/* USER CODE BEGIN PD */

/* USER CODE END PD */

/* Private macro -------------------------------------------------------------*/
/* USER CODE BEGIN PM */

/* USER CODE END PM */

/* Private variables ---------------------------------------------------------*/
ADC_HandleTypeDef hadc1;

I2C_HandleTypeDef hi2c1;

UART_HandleTypeDef huart2;

/* USER CODE BEGIN PV */

/* USER CODE END PV */

/* Private function prototypes -----------------------------------------------*/
void SystemClock_Config(void);
static void MX_GPIO_Init(void);
static void MX_USART2_UART_Init(void);
static void MX_I2C1_Init(void);
static void MX_ADC1_Init(void);
/* USER CODE BEGIN PFP */

/* USER CODE END PFP */

/* Private user code ---------------------------------------------------------*/
/* USER CODE BEGIN 0 */

enum stage { reset, set_temp, watch_temp, buzz };

/* USER CODE END 0 */

/**
  * @brief  The application entry point.
  * @retval int
  */
int main(void)
{
  /* USER CODE BEGIN 1 */
  /* USER CODE END 1 */

  /* MCU Configuration--------------------------------------------------------*/

  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
  HAL_Init();

  /* USER CODE BEGIN Init */
  uint16_t potentiometerValue;
  uint16_t user_temp = 20;
  uint16_t init_temp;
  enum stage curr = reset;
  /* USER CODE END Init */

  /* Configure the system clock */
  SystemClock_Config();

  /* USER CODE BEGIN SysInit */

  /* USER CODE END SysInit */

  /* Initialize all configured peripherals */
  MX_GPIO_Init();
  MX_USART2_UART_Init();
  MX_I2C1_Init();
  MX_ADC1_Init();
  /* USER CODE BEGIN 2 */
  // Display vars below:
  HD44780_Init(2);
  HD44780_Clear();
  HD44780_SetCursor(0,0);
  HD44780_PrintStr("Hi there!");
  HAL_Delay(1700);
  HD44780_Clear();
  /* USER CODE END 2 */

  /* Infinite loop */
  /* USER CODE BEGIN WHILE */
  while (1)
  {
	  if(curr==reset) {
		  if(HAL_GPIO_ReadPin(GPIOA, GPIO_PIN_1) == 0) {
			  curr = set_temp;
			  // Take temperature measurement
			  char i2cdata[2];
			  float temperature;
			  HAL_I2C_Mem_Read(&hi2c1, 0xB4, 0x07, 1, (uint8_t*)i2cdata, 2, 100);
			  int raw_temp = ((i2cdata[1] << 8) | (i2cdata[0]));
			  temperature = raw_temp*0.02 - 273.15;
			  init_temp = temperature;
			  HAL_Delay(10);
			  // Print to display
			  HD44780_Clear();
			  HD44780_SetCursor(0,0);
			  char tempString[10];
			  sprintf(tempString, "%d", (int)temperature);
			  HD44780_PrintStr("Measured temperature:");
			  HD44780_SetCursor(0,1);
			  HD44780_PrintStr(tempString);
			  HAL_Delay(1000);
		  } else {
			  HD44780_Clear();
			  HD44780_SetCursor(0,0);
			  HD44780_PrintStr("Push button");
			  HD44780_SetCursor(0, 1);
			  HD44780_PrintStr("to start");
			  HAL_Delay(1);
		  }
	  } else if(curr==set_temp) {
		  HAL_ADC_Start(&hadc1); // take potentiometer input
		  HAL_ADC_PollForConversion(&hadc1, HAL_MAX_DELAY);
		  potentiometerValue = HAL_ADC_GetValue(&hadc1); // take potentiometer value
		  uint16_t targetTemp = ((potentiometerValue)*(init_temp - 20)/4095)+20; // LINEAR INTERPOLATION FORMULA
		  // Set display params
		  HD44780_Clear();
		  HD44780_SetCursor(0,0);
		  HD44780_PrintStr("Set temperature:"); // print message and temperature
		  char tempString[10];
		  sprintf(tempString, "%hu", targetTemp);
		  HD44780_SetCursor(0,1);
		  HD44780_PrintStr(tempString); // temperature out!
		  HAL_Delay(10);
		  if(HAL_GPIO_ReadPin(GPIOA, GPIO_PIN_1) == 0) {
			  user_temp = targetTemp;
			  curr = watch_temp;
		  }
	  } else if(curr==watch_temp) {
		  // TODO: take temperature measurement -> DONE
		  char i2cdata[2];
		  float temperature;
		  HAL_I2C_Mem_Read(&hi2c1, 0xB4, 0x07, 1, (uint8_t*)i2cdata, 2, 100);
		  int raw_temp = ((i2cdata[1] << 8) | (i2cdata[0]));
		  temperature = raw_temp*0.02 - 273.15;
		  // TODO: print to display -> DONE
		  HD44780_Clear();
		  HD44780_SetCursor(0,0);
		  char tempString[10];
		  sprintf(tempString, "%d", (int)temperature);
		  HD44780_PrintStr("Measured temp:");
		  HD44780_PrintStr(tempString);
		  // TODO: compare with user temp, if close, then BUZZ -> DONE
		  HAL_Delay(1000); // allows user one second before buzz check and activate
		  if(abs(user_temp - temperature)<5) {
			  curr = buzz;
			  continue;
		  } else if(temperature < user_temp) {
			  curr = buzz; // new redundancy feature - if both user_temp and temperature are high but if cup pulled away...
			  continue;
		  }
		  HAL_Delay(9000); // between temp measurements, an overall delay of 10 seconds (subject to change)
	  } else if(curr==buzz) {
		  HAL_GPIO_WritePin(GPIOA, GPIO_PIN_9, GPIO_PIN_SET);
		  HAL_Delay(500);
		  HAL_GPIO_WritePin(GPIOA, GPIO_PIN_9, GPIO_PIN_RESET);
		  // Print display prompt
		  HD44780_Clear();
		  HD44780_SetCursor(0,0);
		  HD44780_PrintStr("Push button");
		  HD44780_SetCursor(0, 1);
		  HD44780_PrintStr("to stop");
		  HAL_Delay(1);
		  if(HAL_GPIO_ReadPin(GPIOA, GPIO_PIN_1) == 0) {
			  curr = reset;
			  HAL_Delay(1000); // bug fix - delay
		  }
	  }
    /* USER CODE END WHILE */
    /* USER CODE BEGIN 3 */
  }
  HD44780_Clear();
  /* USER CODE END 3 */
}

/**
  * @brief System Clock Configuration
  * @retval None
  */
void SystemClock_Config(void)
{
  RCC_OscInitTypeDef RCC_OscInitStruct = {0};
  RCC_ClkInitTypeDef RCC_ClkInitStruct = {0};

  /** Configure the main internal regulator output voltage
  */
  __HAL_RCC_PWR_CLK_ENABLE();
  __HAL_PWR_VOLTAGESCALING_CONFIG(PWR_REGULATOR_VOLTAGE_SCALE2);

  /** Initializes the RCC Oscillators according to the specified parameters
  * in the RCC_OscInitTypeDef structure.
  */
  RCC_OscInitStruct.OscillatorType = RCC_OSCILLATORTYPE_HSI;
  RCC_OscInitStruct.HSIState = RCC_HSI_ON;
  RCC_OscInitStruct.HSICalibrationValue = RCC_HSICALIBRATION_DEFAULT;
  RCC_OscInitStruct.PLL.PLLState = RCC_PLL_ON;
  RCC_OscInitStruct.PLL.PLLSource = RCC_PLLSOURCE_HSI;
  RCC_OscInitStruct.PLL.PLLM = 8;
  RCC_OscInitStruct.PLL.PLLN = 72;
  RCC_OscInitStruct.PLL.PLLP = RCC_PLLP_DIV2;
  RCC_OscInitStruct.PLL.PLLQ = 7;
  if (HAL_RCC_OscConfig(&RCC_OscInitStruct) != HAL_OK)
  {
    Error_Handler();
  }

  /** Initializes the CPU, AHB and APB buses clocks
  */
  RCC_ClkInitStruct.ClockType = RCC_CLOCKTYPE_HCLK|RCC_CLOCKTYPE_SYSCLK
                              |RCC_CLOCKTYPE_PCLK1|RCC_CLOCKTYPE_PCLK2;
  RCC_ClkInitStruct.SYSCLKSource = RCC_SYSCLKSOURCE_PLLCLK;
  RCC_ClkInitStruct.AHBCLKDivider = RCC_SYSCLK_DIV1;
  RCC_ClkInitStruct.APB1CLKDivider = RCC_HCLK_DIV2;
  RCC_ClkInitStruct.APB2CLKDivider = RCC_HCLK_DIV1;

  if (HAL_RCC_ClockConfig(&RCC_ClkInitStruct, FLASH_LATENCY_2) != HAL_OK)
  {
    Error_Handler();
  }
}

/**
  * @brief ADC1 Initialization Function
  * @param None
  * @retval None
  */
static void MX_ADC1_Init(void)
{

  /* USER CODE BEGIN ADC1_Init 0 */

  /* USER CODE END ADC1_Init 0 */

  ADC_ChannelConfTypeDef sConfig = {0};

  /* USER CODE BEGIN ADC1_Init 1 */

  /* USER CODE END ADC1_Init 1 */

  /** Configure the global features of the ADC (Clock, Resolution, Data Alignment and number of conversion)
  */
  hadc1.Instance = ADC1;
  hadc1.Init.ClockPrescaler = ADC_CLOCK_SYNC_PCLK_DIV2;
  hadc1.Init.Resolution = ADC_RESOLUTION_12B;
  hadc1.Init.ScanConvMode = DISABLE;
  hadc1.Init.ContinuousConvMode = DISABLE;
  hadc1.Init.DiscontinuousConvMode = DISABLE;
  hadc1.Init.ExternalTrigConvEdge = ADC_EXTERNALTRIGCONVEDGE_NONE;
  hadc1.Init.ExternalTrigConv = ADC_SOFTWARE_START;
  hadc1.Init.DataAlign = ADC_DATAALIGN_RIGHT;
  hadc1.Init.NbrOfConversion = 1;
  hadc1.Init.DMAContinuousRequests = DISABLE;
  hadc1.Init.EOCSelection = ADC_EOC_SINGLE_CONV;
  if (HAL_ADC_Init(&hadc1) != HAL_OK)
  {
    Error_Handler();
  }

  /** Configure for the selected ADC regular channel its corresponding rank in the sequencer and its sample time.
  */
  sConfig.Channel = ADC_CHANNEL_0;
  sConfig.Rank = 1;
  sConfig.SamplingTime = ADC_SAMPLETIME_3CYCLES;
  if (HAL_ADC_ConfigChannel(&hadc1, &sConfig) != HAL_OK)
  {
    Error_Handler();
  }
  /* USER CODE BEGIN ADC1_Init 2 */

  /* USER CODE END ADC1_Init 2 */

}

/**
  * @brief I2C1 Initialization Function
  * @param None
  * @retval None
  */
static void MX_I2C1_Init(void)
{

  /* USER CODE BEGIN I2C1_Init 0 */

  /* USER CODE END I2C1_Init 0 */

  /* USER CODE BEGIN I2C1_Init 1 */

  /* USER CODE END I2C1_Init 1 */
  hi2c1.Instance = I2C1;
  hi2c1.Init.ClockSpeed = 100000;
  hi2c1.Init.DutyCycle = I2C_DUTYCYCLE_2;
  hi2c1.Init.OwnAddress1 = 0;
  hi2c1.Init.AddressingMode = I2C_ADDRESSINGMODE_7BIT;
  hi2c1.Init.DualAddressMode = I2C_DUALADDRESS_DISABLE;
  hi2c1.Init.OwnAddress2 = 0;
  hi2c1.Init.GeneralCallMode = I2C_GENERALCALL_DISABLE;
  hi2c1.Init.NoStretchMode = I2C_NOSTRETCH_DISABLE;
  if (HAL_I2C_Init(&hi2c1) != HAL_OK)
  {
    Error_Handler();
  }
  /* USER CODE BEGIN I2C1_Init 2 */

  /* USER CODE END I2C1_Init 2 */

}

/**
  * @brief USART2 Initialization Function
  * @param None
  * @retval None
  */
static void MX_USART2_UART_Init(void)
{

  /* USER CODE BEGIN USART2_Init 0 */

  /* USER CODE END USART2_Init 0 */

  /* USER CODE BEGIN USART2_Init 1 */

  /* USER CODE END USART2_Init 1 */
  huart2.Instance = USART2;
  huart2.Init.BaudRate = 115200;
  huart2.Init.WordLength = UART_WORDLENGTH_8B;
  huart2.Init.StopBits = UART_STOPBITS_1;
  huart2.Init.Parity = UART_PARITY_NONE;
  huart2.Init.Mode = UART_MODE_TX_RX;
  huart2.Init.HwFlowCtl = UART_HWCONTROL_NONE;
  huart2.Init.OverSampling = UART_OVERSAMPLING_16;
  if (HAL_UART_Init(&huart2) != HAL_OK)
  {
    Error_Handler();
  }
  /* USER CODE BEGIN USART2_Init 2 */

  /* USER CODE END USART2_Init 2 */

}

/**
  * @brief GPIO Initialization Function
  * @param None
  * @retval None
  */
static void MX_GPIO_Init(void)
{
  GPIO_InitTypeDef GPIO_InitStruct = {0};
/* USER CODE BEGIN MX_GPIO_Init_1 */
/* USER CODE END MX_GPIO_Init_1 */

  /* GPIO Ports Clock Enable */
  __HAL_RCC_GPIOC_CLK_ENABLE();
  __HAL_RCC_GPIOH_CLK_ENABLE();
  __HAL_RCC_GPIOA_CLK_ENABLE();
  __HAL_RCC_GPIOB_CLK_ENABLE();

  /*Configure GPIO pin Output Level */
  HAL_GPIO_WritePin(GPIOA, LD2_Pin|GPIO_PIN_9, GPIO_PIN_RESET);

  /*Configure GPIO pin : PC13 */
  GPIO_InitStruct.Pin = GPIO_PIN_13;
  GPIO_InitStruct.Mode = GPIO_MODE_IT_RISING;
  GPIO_InitStruct.Pull = GPIO_NOPULL;
  HAL_GPIO_Init(GPIOC, &GPIO_InitStruct);

  /*Configure GPIO pin : PA1 */
  GPIO_InitStruct.Pin = GPIO_PIN_1;
  GPIO_InitStruct.Mode = GPIO_MODE_INPUT;
  GPIO_InitStruct.Pull = GPIO_NOPULL;
  HAL_GPIO_Init(GPIOA, &GPIO_InitStruct);

  /*Configure GPIO pins : LD2_Pin PA9 */
  GPIO_InitStruct.Pin = LD2_Pin|GPIO_PIN_9;
  GPIO_InitStruct.Mode = GPIO_MODE_OUTPUT_PP;
  GPIO_InitStruct.Pull = GPIO_NOPULL;
  GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_LOW;
  HAL_GPIO_Init(GPIOA, &GPIO_InitStruct);

/* USER CODE BEGIN MX_GPIO_Init_2 */
/* USER CODE END MX_GPIO_Init_2 */
}

/* USER CODE BEGIN 4 */

/* USER CODE END 4 */

/**
  * @brief  This function is executed in case of error occurrence.
  * @retval None
  */
void Error_Handler(void)
{
  /* USER CODE BEGIN Error_Handler_Debug */
  /* User can add his own implementation to report the HAL error return state */
  __disable_irq();
  while (1)
  {
  }
  /* USER CODE END Error_Handler_Debug */
}

#ifdef  USE_FULL_ASSERT
/**
  * @brief  Reports the name of the source file and the source line number
  *         where the assert_param error has occurred.
  * @param  file: pointer to the source file name
  * @param  line: assert_param error line source number
  * @retval None
  */
void assert_failed(uint8_t *file, uint32_t line)
{
  /* USER CODE BEGIN 6 */
  /* User can add his own implementation to report the file name and line number,
     ex: printf("Wrong parameters value: file %s on line %d\r\n", file, line) */
  /* USER CODE END 6 */
}
#endif /* USE_FULL_ASSERT */
